core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java

```diff
@@ -31,4 +35,46 @@ public void close()
     {
         Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
     }
+
+    public static void withClassLoaderOf(Class<?> clazz, Runnable runnable)
+    {
+        withClassLoader(clazz.getClassLoader(), runnable);
+    }
+
+    public static <T, E extends Exception> T withClassLoaderOf(Class<?> clazz, ThrowingSupplier<T, E> supplier)
+    {
+        return withClassLoader(clazz.getClassLoader(), supplier);
+    }
+
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
+    {
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        catch (RuntimeException e) {
+            throw e;
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
```

Comment:
This is wrong and different behavior than try-with-resources. With the generic parameter for the exception type, you can rethrow the originally. We actually don’t need catch at all — try/finally is sufficient.